L80-csharp-load-balancing.md

```diff
@@ -0,0 +1,387 @@
+        public override Task RefreshAsync(CancellationToken cancellationToken);
+        public override void Start(Action<ResolverResult> listener);
+        protected override void Dispose(bool disposing);
+    }
+    public sealed class DnsResolverFactory : ResolverFactory {
+        public DnsResolverFactory(ILoggerFactory loggerFactory, TimeSpan refreshInterval);
+        public override string Name { get; }
+        public override Resolver Create(Uri address, ResolverOptions options);
+    }
+    public abstract class LoadBalancer : IDisposable {
+        protected LoadBalancer();
+        public void Dispose();
+        public abstract void RequestConnection();
+        public abstract void UpdateChannelState(ChannelState state);
+        protected virtual void Dispose(bool disposing);
+    }
+    public abstract class LoadBalancerFactory {
+        protected LoadBalancerFactory();
+        public abstract string Name { get; }
+        public abstract LoadBalancer Create(IChannelControlHelper controller, IDictionary<string, object> options);
+    }
+    public sealed class PickContext {
+        public PickContext();
+        public HttpRequestMessage? Request { get; set; }
+    }
+    public sealed class PickFirstBalancer : LoadBalancer {
+        public PickFirstBalancer(IChannelControlHelper controller, ILoggerFactory loggerFactory);
+        public override void RequestConnection();
+        public override void UpdateChannelState(ChannelState state);
+        protected override void Dispose(bool disposing);
+    }
+    public sealed class PickFirstBalancerFactory : LoadBalancerFactory {
+        public PickFirstBalancerFactory(ILoggerFactory loggerFactory);
+        public override string Name { get; }
+        public override LoadBalancer Create(IChannelControlHelper controller, IDictionary<string, object> options);
+    }
+    public sealed class PickResult {
+        public Status Status { get; }
+        public Subchannel? Subchannel { get; }
+        public PickResultType Type { get; }
+        public static PickResult ForDrop(Status status);
+        public static PickResult ForFailure(Status status);
+        public static PickResult ForQueue();
+        public static PickResult ForSubchannel(Subchannel subchannel, Action<CompletionContext>? onComplete = null);
+        public void Complete(CompletionContext context);
+    }
+    public abstract class Resolver : IDisposable {
+        protected Resolver();
+        public void Dispose();
+        public abstract Task RefreshAsync(CancellationToken cancellationToken);
+        public abstract void Start(Action<ResolverResult> listener);
+        protected virtual void Dispose(bool disposing);
+    }
+    public abstract class ResolverFactory {
+        protected ResolverFactory();
+        public abstract string Name { get; }
+        public abstract Resolver Create(Uri address, ResolverOptions options);
+    }
+    public sealed class ResolverResult {
```

Comment:
> The ability to indicate an error fetching either the addresses or the service config, independently of each other.

What is the reason for having independent errors? What is each used for?

Reply:
> Attributes associated with each individual address.

Could you give an example of what address attributes are used for? That will help me test that it works.

Reply:
The reason for having independent errors for the addresses and service config is that they can fail independently of each other.  For example, a DNS resolver may succeed in doing an A record lookup to get the addresses, but it may fail doing the TXT lookup for the service config (or vice-versa).  Or the resolver may fetch a service config that it then fails to parse.  If there is a failure getting an updated service config, we don't want that to prevent the channel from getting updated addresses.

Here's what I said in the other thread about examples for using per-address attributes:
> Yes, weighted load balancing is one example. Another example is that when there is a hierarchical tree of LB policies, we can attach attributes to each address to indicate which leaf policy it should be passed to. There are also some other esoteric uses in xDS, like attaching xDS locality names to be used for load reporting purposes.

I'm fairly confident that per-address attributes is something you will need later, even if you don't need it right now.  I'd be okay with not supporting them right now, as long as there's a way to add them later without breaking backward compatibility for this API.

Reply:
I've added both. I'll update the API list in the proposal tomorrow.

Reply:
Updated 😅